project_1/src/simulator.cpp

```diff
@@ -2,60 +2,86 @@
 
 NetworkSimulator::NetworkSimulator(const std::vector<std::uint32_t>& distances_to_receiver) {
     for (std::uint32_t distance: distances_to_receiver) {
-        servers.emplace_back(GenerateNewID(servers), distance);
+        servers.emplace_back(GenerateNewID(servers), distance, packets);
     }
     std::uint32_t server_receiver_id = 0;
     server_receiver = ServerBase(server_receiver_id);
 }
 
-void NetworkSimulator::StartSimulation() {
+void NetworkSimulator::PacketsSending(bool is_first_iteration) {
     // Each server send cwnd_size packets
     for (ServerSender& server: servers) {
-        // Create event
-        events.emplace_back(server, SEND_DATA, server.GetCWNDSize());
-
-        // Create packets, add them to priority_queue
-        for (std::uint32_t i = 0; i < server.GetCWNDSize(); ++i) {
-            // Calculate estimated delivering time of the packet
-            std::uint32_t estimated_delivering_time = server.GetCurrentTime() + server.GetDistance();
-            // Add packet to the priority_queue
-            packets.emplace(server.GetID(), server_receiver.GetID(), server.GetCurrentTime(), i, estimated_delivering_time);
-            
-            // Update server time
-            server.IncreaseCurrentTime(server.GetDistance());
+        // Don't check ACKs on first iteration
+        if (!is_first_iteration && !server.HandleACKs()) {
+            // If server has no ACKs, it doesn't send anything
+            continue;
         }
-        std::cout << events.back();
+        // Create event
+        events.emplace_back(server, event_type::SEND_DATA, server.GetCWNDSize());
+        // Send packets
+        server.SendPackets();
     }
+}
 
+void NetworkSimulator::PacketsReceiving() {
+    std::uint64_t received_data_kb = 0;
+    std::uint32_t received_packets = 0;
     // Create acknowledgements
     while (!packets.empty()) {
+        if (received_data_kb >= bandwidth_bytes) {
```

Comment:
Why do we compare data in kilobytes with data in bytes without any conversion?

Reply:
The same question holds about arithmetic operations.

Reply:
`received_data_kb` -- is a typo, we operate only with bytes.